### ISSUE TITLE
S1 - Newsfeed - Standard Item Container

### ISSUE THREAD

**Issue by jterps08** [PowerlineApp/powerline-rn#13]
See the provided UI. It *might* makes sense to use the FlatList concept as it supports lazy loading, pull to refresh, feed headers, and sticky scrolling. Performance is important. Reference the latest RN capability for lists. ListView is probably going to be deprecated by RN team so FlatList or SectionedList will likely work better.

The Standard Item Container is the container used repeatedly in the newsfeed. It has standard properties for all types of content and it also has special behavior for certain types of content.

1. Avatar (of the author if it's a user post or user petition... of the group if it's a group poll/petition/discussion/fundraiser/event... uses imgix).. should hyperlink to the appropriate user/group profile screen
2. Author Name (the author who created the item)... should hyperlink to the author's user profile screen
3. Group Name (where it was posted)... should hyperlink to the group profile screen 
4. Text Body (main body of item, up to 300 characters)... should hyperlink to the Item Detail screen unless its a URL, an @Mention, or a hashtag. We will address @mentions, hashtags, and URL Metadata preview as separate issues.
5. Item age (how old it is)
6. Engagement counter (provides vote total + comments total)
7. A drop-down menu (for secondary functions as seen on mockup)
8. "Quick Action Buttons" (easy ability to upvote the post directly from the newsfeed, etc.)

Special behavior: 
1. If the item is a petition/fundraiser/event, it will have a Title. This title should display as seen in the mockup as bolded text above the standard body area. Contact me for clarification on exact position if unclear
1. If item is priority zone, it displays the priority zone icon (lightning icon) as seen in the provided UI.
2. If the item has an attachment (picture or video), the first picture or video is displayed beneath the item.
3. If the user post or user petition has a URL in the body, the URL Metadata is displayed beneath the item (image, title, description). If the backend provides this, it should be displayed on a user post or user petition.
4. If item is user post, it displays Upvote and Downvote buttons (see mockup). Tapping Upvote or Downvote actually upvotes/downvotes the post and does not open the item detail screen. 
5. If item is user petition, it displays "Sign" button instead of Upvote/Downvote buttons. Tapping Sign actually signs the petition and does not open the item detail screen. The button changes to "Unsign"
6. If item is group poll, it displays "Answer" instead of Upvote/Downvote buttons. Tapping Answer opens the item detail screen.
7. If item is group petition, it displays "Sign" instead of Upvote/Downvote buttons. Tapping Sign actually signs the petition and does not open the item detail screen.
8. If item is group fundraiser, it displays "Pay" instead of Upvote/Downvote buttons. Tapping Pay opens the item detail screen.
9. If item is group event, it displays "RSVP" instead of Upvote/Downvote buttons. Tapping RSVP opens the item detail screen.
10. If item is group discussion, it displays "Discuss" instead of Upvote/Downvote buttons. Tapping discuss opens the item detail screen.
11. If there are any comments, the app will display a preview of the first most recent comment below the item. Tapping on this 'comment preview' will open the Item Detail Screen. 

General notes
- Unless otherwise specified above (e.g avatars, group names, user names, hashtags, URLs), tapping on the standard item container should open up the Item Detail Screen. Tapping the Reply button should open up the Reply Modal.
- (out of scope for this issue) Imgix processing should be used for image compression on-the-fly. See v2.0 for current functionality. #40
- (out of scope for this issue) Tapping a picture attachment opens the picture in full-screen view. The Youtube video attachment should be able to be played in-line #41

**Comment by jterps08** [comment]
Be sure to leverage NativeBase Card - http://docs.nativebase.io/Components.html#card-def-headref
Standard Item Container will be filled by `Activity` in Newsfeeds... will be filled by `Bookmarks` for Favorites

**Comment by igaponov** [comment]
https://api-dev.powerli.ne/api-doc#get--api-v2-activities to get activities
1. attribute `zone`
3. only for polls, attribute `educational_context`
4. attr `metadata`
12. last comment can be retrieved from `GET  /api/v2/[entity type]/{id}/comments`, sort by createdAt DESC, but I guess it's not a desirable behavior as frontend will have to make many requests. At the other hand, the query for newsfeed is overcomplicated now and adding three new joins is not a good idea. @jterps08 what do you think about it?

**Comment by jterps08** [comment]
@igaponov I see how item 12 will be problematic... We haven't done query optimization for newsfeed yet, right? Would it make sense to do query optimization for newsfeed so that we can optimize query optimization? Or alternatively @austinpapp do you have any ideas?

**Comment by igaponov** [comment]
@jterps08 we can't optimize this query. If we want to add comments we should add joins. The only thing we can do here - make the same thing we did for reports (add a table with denormalized data).

**Comment by jterps08** [comment]
I need @austinpapp to weigh in here. At this point, I am thinking we do the joins and refactor later, but I really don't know the best path. I just want the endpoint to be ready for Awis so that he can leverage it. Even if endpoint is static (and always provides fake stuff back) in the short term so that he has an endpoint to work with, I think that would be helpful while we figure out the answer for this

**Comment by jterps08** [comment]
Reopening... We are missing #12 in the list above on the Standard Item Container... 

> If there are any comments, the app will display a preview of the first most recent comment below the item. Tapping on this 'comment preview' will open the Item Detail Screen.

![image](https://user-images.githubusercontent.com/7582457/29872395-734880d6-8d5d-11e7-8763-9544a2dab2e8.png)

**Comment by jterps08** [comment]
We are also missing item 4 URL Metadata Preview... I saw this working, so I don't know what happened https://github.com/PowerlineApp/powerline-rn/commit/7a2c5177b5ffd0e3f00322fbc7ba9ce9094fd501

**Comment by jterps08** [comment]
After #128

**Comment by jterps08** [comment]
@piubellofelipe I noticed this on Android. You shoudl look at this for 13>>>

![image](https://user-images.githubusercontent.com/7582457/31751186-1a102d86-b452-11e7-951b-69272b173ad5.png)

**Comment by jterps08** [comment]
Closing, but I still don't see metadata correctly. assuming this is a backend issue.

**Comment by jterps08** [comment]
Comment Preview works. Attachments works. Metadata does not work.

**Comment by piubellofelipe** [comment]
via Hubstaff
User: Felipe Piubello

Project: Powerline React Native - https://app.hubstaff.com/projects/264175
Date Range: 10/10/17 - 10/22/17
Work session total: 7:54:44

Grand total: 7:54:44
